lib/util/map.ts

```diff
@@ -0,0 +1,65 @@
+/// <reference path='../../typings/tsd.d.ts' />
+
+import _ = require('lodash');
+
+export = Map;
+
+class Map<T> {
+    private _size: number = 0;
+    private _map: {[key: string]: T} = {};
+
+    get size(): number {
+        return this._size;
+    }
+
+    clear(): void {
+        this._size = 0;
+        this._map = {};
+    }
+
+    set(key: string|number, val: T): void {
+        if (!_.has(this._map, key.toString())) {
```

Comment:
Let's cache the result of `.toString()` since it is used twice in this function.